hadoop-hdds/docs/content/design/storage-class.md

```diff
@@ -19,10 +20,408 @@ author: Marton Ele
+
+## Multi-stage storage-class
+
+There is a very specific use case the "temperature" of data. Key (and containers) can become COLD or HOT over the the time. As we need transitions between the different **state** of the containers, the COLD data should be a new state of a container not a storage-class.
+
+COLD and HOT data are not planned to be added, but a good example for the abstraction level of storage-classes.These are not different replication mechanism, but possible new states in an existing replication workflow.
+
+![Storage class transitions](storage-class-transition.png)
+
+# Implementation changes
+
+The goal with the implementation is to introduce a new abstraction level, but use exactly the same replication mechanism as today.
+
+
+**New interfaces**: To manage the storage-class interfaces we need to define the interfaces:
+
+ * We need configuration classes for the existing replication mechanism to define the replication factor.
+    * For normal `RATIS` replication only the `ratis_quorum`, and the `required_closed_replicas` can be configured and only two tuples are valid `(ONE,ONE)`, `(THREE,THREE)`
+    * `STANDALONE` replication is deprecated but we need to support it until full removal (without parameters)
+ * `StorageClass` is a POJO with the selection of `ReplicationMechanism` and the replication configuration.
+ * We need a `StorageClassRegistry` instance which is available both from client side and server side. Today we don't need to make the configurable from external files: both the existing replication mechanisms can be configured from the code. 
+ * `StorageClassRegistry` will contain the definition of the three initial storage-classes: STANDARD (RATIS/THREE), REDUCED (RATIS/ONE) and LEGACY (STANDALONE)
+ * We need a *legacy mapping table* which maps the `factor/type` from any older requests to `storageClass`! This can help to keep the backward compatibility.
+  
+**Protocol changes**: 
+
+  * `storageClass` field should be added to all the protocols where we use `factor/type` today. Old `factor/type` fields will be kept for compatibility reasons. 
+  * Storage-class will be represented as a custom String in the RPC protocol (in storage-sensitive places it can be stored as an integer with an additional mapping).
+  * There are two protocols which should be changed:
+   * *Key creation* and *Key information* related structures in `OmClientProtocol` should be updated (`KeyInfo`, `MultipartUploadInfo`, `KeyArgs`,... )
+   *  SCMAdmin protocol should be improved to include `storage-class` in `ContainerRequestProto`.
+
+**Ozone Client changes**: Ozone client classes (`OzoneBucket`, `OzoneKey`,...) should be modified to propagate the `storageClass` everywhere instead of `factor/type`. To make it backward compatible we can create legacy mapping table should be used.
```

Comment:
This one sentence cannot capture all the client changes right? The client has to select different data paths depending on the replication scheme (Ratis vs EC). How will that be abstracted out?

Reply:
Thanks the question. The implementation details of EC is intentionally removed from this doc as you suggested. This document focuses on the current replication schemes. But EC can be handed in the same way:

The current implementation of OzoneClient calls OM first to create a new key, and uses the returned pipeline information to initialize the right `XCeiverClient` (`XCeiverClientManager.getClient`)

With storage-class the logic is the same, the only required change is adding storage-class to the `CreateKey` request and use it instead of replication/factor. The storage-class will be used in the `AllocateScmBlock` call, and SCM will return with the right pipeline.

But this is an SCM change, client-side it's enough to have the mentioned  change.  

In fact if we implement it once, we don't need to modify the logic for EC, just implement a new `XCeiverClientSpi`. Without storage-class we need to modify the `CreateKey` (or at least adding a new replication type, which doesn't use the factor at all.)